### ISSUE TITLE
UnicodeDecodeError: 'charmap' codec can't decode byte 0x8d in position 2869: character maps to <undefined>

### ISSUE THREAD

**Issue by Weilin37** [MilaNLProc/contextualized-topic-models#25]
* Contextualized Topic Models version: 1.5.2
* Python version: 3.7
* Operating System: Windows 10

### Description

I'm just copy and pasting the notebook found here, with the difference that instead of using wget on the dataset, I just downloaded it from the colab notebook: 

https://colab.research.google.com/drive/1V0tkpJL1yhiHZUJ_vwQRu6I7_svjw1wb?usp=sharing#scrollTo=WXZ8fOdYwdWO

### What I Did

I then get the following error.
```
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-8-5453e403b509> in <module>
      3         training_bert = pickle.load(filino)
      4 else:
----> 5     training_bert = bert_embeddings_from_file("dbpedia_sample_abstract_20k_unprep.txt", "distilbert-base-nli-mean-tokens", batch_size=50)
      6     with open("saved_embeddings", "wb") as filino:
      7         pickle.dump(training_bert, filino)

~\AppData\Roaming\Python\Python38\site-packages\contextualized_topic_models\utils\data_preparation.py in bert_embeddings_from_file(text_file, sbert_model_to_load, batch_size)
     21     model = SentenceTransformer(sbert_model_to_load)
     22     with open(text_file) as filino:
---> 23         train_text = list(map(lambda x: x, filino.readlines()))
     24 
     25     return np.array(model.encode(train_text, show_progress_bar=True, batch_size=batch_size))

c:\program files\python38\lib\encodings\cp1252.py in decode(self, input, final)
     21 class IncrementalDecoder(codecs.IncrementalDecoder):
     22     def decode(self, input, final=False):
---> 23         return codecs.charmap_decode(input,self.errors,decoding_table)[0]
     24 
     25 class StreamWriter(Codec,codecs.StreamWriter):

UnicodeDecodeError: 'charmap' codec can't decode byte 0x8d in position 2869: character maps to <undefined>
```

**Comment by silviatti** [comment]
Hello!

The problem seems to be related to the encoding in Windows: the default encoding type is not "utf-8" (I guess [this](https://www.python.org/dev/peps/pep-0597/) is related. We will try to add support for it in the next few days.

In the meantime, instead of using the method `bert_embeddings_from_file`, you can use the method `bert_embeddings_from_list` ([link to code](https://github.com/MilaNLProc/contextualized-topic-models/blob/master/contextualized_topic_models/utils/data_preparation.py#L28)) that takes as input a list of strings (the documents to convert in embeddings). 

In other words, you replace this line :
```python
training_bert = bert_embeddings_from_file("dbpedia_sample_abstract_20k_unprep.txt", "distilbert-base-nli-mean-tokens", batch_size=50)
```

with these lines:
```python
with open(your_text_file, "utf-8") as filino:
      sentences = list(map(lambda x : x.replace("\n", ""), filino.readlines())
training_bert = bert_embeddings_from_list(sentences "distilbert-base-nli-mean-tokens", batch_size=50)
```

This should fix your problem :)

Silvia

**Comment by vinid** [comment]
Hi @Weilin37,

In the meantime, I have patched the package and added the utf-8 default encoding to the `with open` statements. If you install the new version with `pip install -U` you should be able to run everything.

Let me know if it works :)

**Comment by vinid** [comment]
closing this issue, for now, feel free to ping us if you need help

**Comment by Weilin37** [comment]
Thanks! I haven't time to try it yet but if it doesn't work I will update

**Comment by Weilin37** [comment]
update: Got this new error

```
 File "<ipython-input-6-88616fd02b18>", line 2
    with open("saved_embeddings", "rb"') as filino:
                                                   ^
SyntaxError: EOL while scanning string literal
```

Any idea what that means?

**Comment by vinid** [comment]
I see a ' char that generates the error.

wrong:  `with open("saved_embeddings", "rb"') as filino:`
good: `with open("saved_embeddings", "rb") as filino:`

where did you find this error? I might have left this thing somewhere and I have to correct it :)

**Comment by Weilin37** [comment]
It works! thank you. Generally speaking, how am I able to pair the topics found to each document? Is there a way to link the results back to the original dataset?

**Comment by vinid** [comment]
You can look at this [colab notebook](https://colab.research.google.com/drive/1quD11TMy-1D-GxPUj_Dea4iRYmOO8C2C?usp=sharing)

Essentially, this is the code you need:

``` python
topics_predictions = ctm.get_thetas(training_dataset, n_samples=20) # get all the topic predictions
print(preprocessed_documents[id_of_the_document_you_are_interested_in]) # see the text of our preprocessed document

import numpy as np
topic_number = np.argmax(topics_predictions[id_of_the_document_you_are_interested_in]) # get the topic id of the first document
ctm.get_topic_lists(5)[topic_number] #and the topic should be about natural location related things ```

`ctm.get_thetas` allows you to get the topic distribution
`np.argmax` will tell you the index of the topic that is the most probable for a document
